sys/vane/clay.hoon

```diff
@@ -402,181 +426,1362 @@
+          ::
+          ?:  =/  target-mark=mark  =+(spur=(flop pax) ?~(spur !! i.spur))
+              ?.  =(target-mark p.cag)
+                %.n
+              ::
+              =/  stored  (need (need (read-x:sutil & let.dom pax)))
+              =/  stored-cage=cage  ?>(?=(%& -.stored) p.stored)
+              ::
+              =(q.q.stored-cage q.q.cag)
+            ::
+            $(soba t.soba)
+          ::  the value differs from what's stored, so register mutation
+          ::
+          $(soba t.soba, mut.nuz [i.soba mut.nuz])
+        ==
+      ::  sort each section alphabetically for determinism
       ::
-      :-  `hun.rom.ruf
-      =/  jod  (fall (~(get by dos.rom.ruf) syd) *dojo)
-      [lim=now ref=~ [qyx dom dok mer per pew]:jod]
-  ::
-  =*  red=rede  ->
-  =|  mow/(list move)
-  |%
-  ++  abet                                              ::  resolve
-    ^-  [(list move) raft]
-    :-  (flop mow)
-    ?.  =(our her)
-      ::  save foreign +rede
+      =.  nuz  :*
+        (sort del.nuz sort-by-head)
+        (sort ins.nuz sort-by-head)
+        (sort dif.nuz sort-by-head)
+        (sort mut.nuz sort-by-head)
+        (sort ink.nuz sort-by-head)
+      ==
+      =/  ink
+         %+  turn  ink.nuz
+         |=  {pax/path mis/miso}
+         ^-  (pair path cage)
+         ?>  ?=($ins -.mis)
+         =+  =>((flop pax) ?~(. %$ i))
+         [pax - [%atom %t ~] ((hard @t) +>.q.q.p.mis)]
       ::
-      =/  run  (fall (~(get by hoy.ruf) her) *rung)
-      =?  rit.run  =(0 rit.run)
-        (fall (rift-scry her) *rift)
-      =/  rug  (~(put by rus.run) syd red)
-      ruf(hoy (~(put by hoy.ruf) her run(rus rug)))
-    ::  save domestic +room
+      =/  mim
+        ::  add the new files to the new mime cache
+        ::
+        %-  malt
+        ^-  (list (pair path mime))
+        ;:  weld
+          ^-  (list (pair path mime))
+          %+  murn  ins.nuz
+          |=  {pax/path mis/miso}
+          ^-  (unit (pair path mime))
+          ?>  ?=($ins -.mis)
+          ?.  ?=($mime p.p.mis)
+            ~
+          `[pax ((hard mime) q.q.p.mis)]
+        ::
+          ^-  (list (pair path mime))
+          %+  murn  ink.nuz
+          |=  {pax/path mis/miso}
+          ^-  (unit (pair path mime))
+          ?>  ?=($ins -.mis)
+          ?>  ?=($mime p.p.mis)
+          `[pax ((hard mime) q.q.p.mis)]
+        ::
+          ^-  (list (pair path mime))
+          %+  murn  mut.nuz
+          |=  {pax/path mis/miso}
+          ^-  (unit (pair path mime))
+          ?>  ?=($mut -.mis)
+          ?.  ?=($mime p.p.mis)
+            ~
+          `[pax ((hard mime) q.q.p.mis)]
+        ==
+      ::
+      ;<  ins=(list (pair path cage))       bind:m  (calc-inserts wen ins.nuz)
+      ;<  dif=(list (trel path lobe cage))  bind:m  (calc-diffs wen dif.nuz)
+      ;<  mut=(list (trel path lobe cage))  bind:m  (calc-mutates wen mut.nuz)
+      %+  pure:m
+        ^-  dork
+        [del=(turn del.nuz head) ink ins dif mut]
+      mim
```

Comment:
This looks like it modifies the mime cache only after all the Ford operations complete successfully. We calculate the mime cache eagerly, though, so I'm not sure we should delay applying it. It does seem a litle weird to prime the cache with values that eventually don't validate, but in that case I feel like we should only perform the coercion to mime after the Ford builds finish, so the errors will happen quicker.

Maybe it's not a big deal either way; curious what you think about this.

Reply:
I changed the behavior of the mime cache somewhat.  Before, it did two things:

- When we get a sync from unix (where everything comes in marked `%mime`), store the raw `%mime` version so that at the end of the commit we can just spit those back to unix rather than having to recompute them all.
- When we get a sync from unix, check whether the value is the same as what's in the mime cache (which was updated the last time we got a sync from unix).  There was a bug here in that once a file was changed from unix, then from within arvo, the arvo change wouldn't be synced to unix (because the cache only got updated when writing from unix).  @Fang- you may remember this bug from the azimuth deploy.  This was a significant bug that really contributed to my feeling of Clay being unstable because it felt like Clay got "unsynced" from Unix.

Now it only does the first of these.  This simplifies our state because the mime cache now only lasts for a single transaction, so it's not in our permanent state.  The cost is that we have to convert from `%mime` more often, but in practice it seems that the mime conversion hits ford's cache and doesn't cause a problem even for my "copy everything in every 2 seconds" workflow, which is close to the pathological case.

This also fixes the bug I mentioned.  If you really wanted to preserve that mime cache for the second usage, I think you could do it by making sure to update the mime cache every time you write to unix as well as when you read.  With Ford's good behavior, I figured it wasn't worth the complexity.

Reply:
I remember that bug! +1 for it being gone. (: